File: 2kyu/esolang_interpreters.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <criterion/criterion.h>

#define BUFSIZE 30000

typedef struct {
  char *data;
  int size;
} Bits;

Bits* bits_alloc() {
  Bits *bits = (Bits*)malloc(sizeof(Bits));
  bits->size = 0;
  bits->data = NULL;
  return bits;
}

Bits* append(Bits *bits, char ch) {
  assert(bits);
  bits->size++;
  if (bits->data == NULL) {
    bits->data = (char*)malloc(bits->size);
  } else {
    bits->data = (char*)realloc(bits->data, bits->size);
  }
  bits->data[bits->size - 1] = ch;
  return bits;
}

Bits* string_to_bits(char *in) {
  Bits *bits = (Bits*)malloc(sizeof(Bits));  
  bits->size = strlen(in) * 8;
  bits->data = (char*)malloc(bits->size);
  memset(bits->data, 0, bits->size);  
  for (int i = 0; *in != '\0'; i += 8, in++) {
    unsigned char ch = *in;
    int pos = i;
    while (ch > 0) {
      bits->data[pos++] = ch % 2;
      ch /= 2;
    }
  }
  return bits;
}

char fetch_bit(Bits *bits, int ndx) {
  if (ndx >= bits->size || ndx < 0) return 0;
  return bits->data[ndx];
}

char byte_to_char(char *bits, int start, int end) {
  assert(bits);
  int sum = 0;
  int prod = 1;
  const int base = 2;
  for (int i = start; i < end; i++) {
    sum += prod * bits[i];
    prod *= base;
  }
  return sum;
}

char* bits_to_string(Bits *bits) {
  assert(bits);
  int size = 1;
  char *s = (char*)malloc(size);
  for (int i = 0; i <= bits->size; i += 8) {
    int end = i + 8 > bits->size ? bits->size : i + 8;
    s[size-1] = byte_to_char(bits->data, i, end);
    size++;
    s = (char*)realloc(s, size);
  }
  s[size-1] = '\0';
  return s;
}

char* boolfuck (char *code, char *in) {
  char data[BUFSIZE];
  memset(data, 0, sizeof(data));
  int dp = BUFSIZE / 2;
  Bits *tape = string_to_bits(in);
  Bits *bits = bits_alloc();
  int bp = 0;
  while (*code != '\0') {
    switch (*code) {
    case '+':
      data[dp] = !data[dp];
      break;
    case ',':
      data[dp] = fetch_bit(tape, bp++);
      break;
    case ';':
      bits = append(bits, data[dp]);
      break;
    case '>':
      dp++;
      break;
    case '<':
      dp--;
      break;
    case '[':
      if (!data[dp]) {
        int ct = 0;
        while (!(*code == ']' && ct == 1)) {
          if (*code == ']') ct--;
          else if (*code == '[') ct++;
          code++;
        }
      }
      break;
    case ']':
      if (data[dp]) {
        int ct = 0;
        while (!(*code == '[' && ct == 1)) {
          if (*code == ']') ct++;
          else if (*code == '[') ct--;
          code--;
        }
      }
      break;
    }
    code++;
  }

  char *s = bits_to_string(bits);
  free(tape);
  free(bits);
  return s;
}

File: 2kyu/esolang_interpreters.py
N = 30000

def byteToChar(bits):
    su = 0
    prd = 1
    for b in bits:
        su += prd * b
        prd *= 2
    return chr(su)

class Bits(object):
    def __init__(self):
        self._size = 0;
        self._data = []

    def append(self, ch):
        self._data.append(ch)
        self._size += 1

    def fromString(self, s):
        self._size = len(s) * 8
        self._data = [0] * self._size
        i = 0
        for c in s:
            n = ord(c)
            pos = i
            while n > 0:                
                self._data[pos] = n % 2
                n //= 2
                pos += 1
            i += 8
    def __str__(self):
        s = ''
        for i in range(0, self._size, 8):
            s += byteToChar(self._data[i:i+8])
        return s

    def __getitem__(self, key):
        if key >= self._size: return 0
        return self._data[key]
    
    __repr__ = __str__
    
def boolfuck(code, inpt = ""):
    data = [0] * N
    dp = N // 2

    tape = Bits()
    tape.fromString(inpt)
    tp = 0
    
    bits = Bits()

    clen = len(code)
    cp = 0;
    while (cp < clen):
        c = code[cp]
        if c == '+':
            data[dp] = (data[dp] + 1) % 2
        elif c == ',':
            data[dp] = tape[tp]
            tp += 1
        elif c == ';':
            bits.append(data[dp])
        elif c == '>':
            dp += 1
        elif c == '<':
            dp -= 1
        elif c == '[':
            if data[dp] == 0:
                ct = 0
                while not (code[cp] == ']' and ct == 1):
                    if code[cp] == ']':
                        ct -= 1
                    elif code[cp] == '[':
                        ct += 1
                    cp += 1
        elif c == ']':
            if data[dp] != 0:
                ct = 0
                while not (code[cp] == '[' and ct == 1):
                    if code[cp] == ']':
                        ct += 1
                    elif code[cp] == '[':
                        ct -= 1
                    cp -= 1        
                
        cp += 1
        
    return str(bits)
